### PR TITLE
Manifest file creation functionality added in data egress

### PR DESCRIPTION
**Pull request by umairpgh**
Manifest file creation functionality added in data egress

### CODE REVIEW

File: src/main/kotlin/uk/gov/dwp/dataworks/egress/services/impl/DataServiceImpl.kt [hunk 1]
```diff
@@ -58,6 +59,10 @@ class DataServiceImpl(
         if (specification.controlFilePrefix?.isNotBlank() == true) {
             controlFileService.egressControlFile(egressed, specification)
         }
+
+        if (specification.manifestFileName?.isNotBlank() == true) {
+            manifestFileService.egressManifestFile(specification)
```
Comment (rbryett) [inline src/main/kotlin/uk/gov/dwp/dataworks/egress/services/impl/DataServiceImpl.kt:64]: We call egressManifestFile, but don't use the Pair it returns. Is that right?

File: src/test/kotlin/uk/gov/dwp/dataworks/egress/services/impl/ManifestFileServiceImplTest.kt [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+import java.io.File
+import java.text.DecimalFormat
+import java.text.SimpleDateFormat
+import java.util.*
+import java.util.concurrent.CompletableFuture
+
+class ManifestFileServiceImplTest: StringSpec() {
+
+    init {
+            "Export manifest file as plan text if no manifestFileEncryption option provided" {
```
Comment (rbryett) [inline src/test/kotlin/uk/gov/dwp/dataworks/egress/services/impl/ManifestFileServiceImplTest.kt:31]: typo, should be "plain text"

File: src/main/kotlin/uk/gov/dwp/dataworks/egress/services/impl/ManifestFileServiceImpl.kt [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+
+        val sourcePrefix = specification.sourcePrefix
+        val todaysDate = if (sourcePrefix.contains(TODAYS_YYYYMMDD_FORMATTED_DATE_PLACEHOLDER)) todaysDate("yyyyMMdd")
+                            else todaysDate()
+
+        // DataProductType only applicable for HTME Collections, or anything containing full or incremental in source prefix, last - is removed
+        val dataProductType = if (sourcePrefix.lowercase().contains("full")) "full"
+                                else if (sourcePrefix.lowercase().contains("incremental")) "incremental"
+                                else ""
+        // DataProduct is mane of HTME Collections, taken from source prefix and last - is removed if present
```
Comment (rbryett) [inline src/main/kotlin/uk/gov/dwp/dataworks/egress/services/impl/ManifestFileServiceImpl.kt:96]: typo "mane"

File: src/main/kotlin/uk/gov/dwp/dataworks/egress/services/impl/ManifestFileServiceImpl.kt [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+    }
+
+    private suspend fun targetContentsAndRequest(specification: EgressSpecification, targetKey:String): Pair<ByteArray, PutObjectRequest> {
+
+        val sourcePrefix = specification.sourcePrefix
+        val todaysDate = if (sourcePrefix.contains(TODAYS_YYYYMMDD_FORMATTED_DATE_PLACEHOLDER)) todaysDate("yyyyMMdd")
+                            else todaysDate()
+
+        // DataProductType only applicable for HTME Collections, or anything containing full or incremental in source prefix, last - is removed
+        val dataProductType = if (sourcePrefix.lowercase().contains("full")) "full"
```
Comment (rbryett) [inline src/main/kotlin/uk/gov/dwp/dataworks/egress/services/impl/ManifestFileServiceImpl.kt:93]: Can't we just set dataProductType to the value? i.e.
val dataProductType = sourcePrefix.lowercase()

Comment (umairpgh) [reply src/main/kotlin/uk/gov/dwp/dataworks/egress/services/impl/ManifestFileServiceImpl.kt:93]: source prefix is full path to the location of that collection which contains full or incremental as substring

File: src/main/kotlin/uk/gov/dwp/dataworks/egress/services/impl/DataServiceImpl.kt (hunk 1 above)
Comment (umairpgh) [reply src/main/kotlin/uk/gov/dwp/dataworks/egress/services/impl/DataServiceImpl.kt:64]: Yes, we don't need further to use these values, the service fails if it couldn't process manifest file. So no further action to be taken by DataService, the returning pair  is following the convention used by ControlFile.

File: src/test/kotlin/uk/gov/dwp/dataworks/egress/services/impl/ManifestFileServiceImplTest.kt (hunk 2 above)
Comment (umairpgh) [reply src/test/kotlin/uk/gov/dwp/dataworks/egress/services/impl/ManifestFileServiceImplTest.kt:31]: updated

File: src/main/kotlin/uk/gov/dwp/dataworks/egress/services/impl/DataServiceImpl.kt (hunk 1 above)
Comment (umairpgh) [reply src/main/kotlin/uk/gov/dwp/dataworks/egress/services/impl/DataServiceImpl.kt:64]: The returning pair of manifestFileService is the following the convention of controlFileService. The service fails if manifest file is not processed successfully, so no further action is taken outside of the manifestFileService.

Comment (rbryett) [reply src/main/kotlin/uk/gov/dwp/dataworks/egress/services/impl/DataServiceImpl.kt:64]: Given we don't implement the controlFileService interface, do we need to follow the convention? I can't see where we use the returned Pair, unless im missing something?

**Review by rbryett — APPROVED** [review]

**Pull request merged**